### PR TITLE
Add api endpoints santa rules

### PR DESCRIPTION
**Pull request by xarbit**

### CODE REVIEW

File: zentral/contrib/santa/serializers.py [hunk 1]
```diff
@@ -59,18 +62,264 @@ def update(self, instance, validated_data):
...
+            primary_users=validated_data.get("primary_users") or [],
+            excluded_primary_users=validated_data.get("excluded_primary_users") or []
+        )
+        tags = validated_data.get("tags")
+        if tags:
+            rule.tags.set(tags)
+        excluded_tags = validated_data.get("excluded_tags")
+        if excluded_tags:
+            rule.excluded_tags.set(excluded_tags)
+        return rule
```
Comment (np5) [inline zentral/contrib/santa/serializers.py:193]: If you want to add the events in this PR, you should also post an event here. see https://github.com/zentralopensource/zentral/blob/d301166954f7b754d90252791afb2bdca1ef4bcf/zentral/contrib/santa/views/setup.py#L294-L295

File: tests/santa/test_api_views.py [hunk 2]
```diff
@@ -713,6 +723,632 @@ def test_rule_list_by_unknown_configuration(self):
...
+            "target_identifier": rule.target.identifier,
+            "description": "Description Text Updated"
+        }
+        response = self.put_json_data(reverse("santa_api:rule", args=(rule.pk,)), data)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        rule.refresh_from_db()
+        self.assertEqual(response.json()["description"], "Description Text Updated")
+        self.assertEqual(rule.description, "Description Text Updated")
+
+    def test_update_rule_change_target_identifier_error(self):
```
Comment (np5) [inline tests/santa/test_api_views.py:1035]: It should be possible to change a rule target

File: tests/santa/test_api_views.py [hunk 3]
```diff
@@ -713,6 +723,632 @@ def test_rule_list_by_unknown_configuration(self):
...
+            "target_type": rule.target.type,
+            "target_identifier": get_random_string(length=64, allowed_chars='abcdef0123456789')
+        }
+        response = self.put_json_data(reverse("santa_api:rule", args=(rule.pk,)), data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.json(), {
+            "target": ["does not exist!"]
+        })
+
+    def test_update_rule_change_target_type_error(self):
```
Comment (np5) [inline tests/santa/test_api_views.py:1051]: It should be possible to change a rule target

File: tests/santa/test_api_views.py [hunk 4]
```diff
@@ -713,6 +723,632 @@ def test_rule_list_by_unknown_configuration(self):
...
+        rule = self.force_rule(configuration=configuration)
+        response = self.put_json_data(reverse("santa_api:rule", args=(rule.pk,)), {})
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_update_rule_unauthorized(self):
+        configuration = self.force_configuration()
+        rule = self.force_rule(configuration=configuration)
+        response = self.put_json_data(reverse("santa_api:rule", args=(rule.pk,)), {}, include_token=False)
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
```
Comment (np5) [inline tests/santa/test_api_views.py:1322]: You should add tests to trigger all the conditions in the building of the update rule event.

**Review by np5 — CHANGES_REQUESTED** [review]

File: zentral/contrib/santa/serializers.py [hunk 5]
```diff
@@ -59,18 +62,264 @@ def update(self, instance, validated_data):
         return super().update(instance, validated_data)
 
 
-class RuleTargetSerializer(serializers.ModelSerializer):
-    class Meta:
-        model = Target
-        fields = ("type", "identifier")
+class RuleMixin:
```
[the remark this answers is not in the document]
Comment (xarbit) [reply zentral/contrib/santa/serializers.py:65]: Thx. Yes, wanted to change that today, that is why I marked it as "Draft"

File: zentral/contrib/santa/serializers.py [hunk 6]
```diff
@@ -59,18 +62,118 @@ def update(self, instance, validated_data):
...
+        return super().create(validated_data)
+
+    def update(self, instance, validated_data):
+        target, _ = Target.objects.update_or_create(
+            type=validated_data.pop("target_type"),
+            identifier=validated_data.pop("target_identifier")
+        )
+        validated_data["target"] = target
+        if validated_data.get("custom_msg") != instance.custom_msg:
+            instance.version = F("version") + 1
```
Comment (np5) [inline zentral/contrib/santa/serializers.py:173]: You should bump the version if any one of the attribute is changed.

**Review by np5 — CHANGES_REQUESTED** [review]

**Review by np5 — APPROVED** [review]

**Pull request merged**